orchagent/aclorch.cpp

```diff
@@ -3903,10 +4228,194 @@ bool AclOrch::updateAclTable(AclTable &currentTable, AclTable &newTable)
+            }
+            egrSetDscpTable.link(bind_port_id);
+            egrSetDscpTable.portSet.emplace(alias);
+        }
+
+        egrSetDscpTable.id = EGR_SET_DSCP_TABLE_ID;
+        egrSetDscpTable.stage = ACL_STAGE_EGRESS;
+        auto egrSetDscpTableType = getAclTableType(TABLE_TYPE_EGR_SET_DSCP);
+        sai_object_id_t egrSetDscp_oid = getTableById(EGR_SET_DSCP_TABLE_ID);
+        // create the EGR_SET_DSCP fisrt time if not present. Otherwise update the existing table.
+        if (m_egrSetDscpRef.empty())
+        {
+            // Create EGR_SET_DSCP table
+            egrSetDscpTable.validateAddType(*egrSetDscpTableType);
+            egrSetDscpTable.addMandatoryActions();
+            if (!egrSetDscpTable.validate())
+            {
+                SWSS_LOG_ERROR("Failed to validate ACL table %s",
+                    EGR_SET_DSCP_TABLE_ID);
+                // since we failed to create the table, there is no need for rollback.
+                return EgressSetDscpTableStatus::EGRESS_SET_DSCP_TABLE_FAILED;
+            }
+            if (!addAclTable(egrSetDscpTable))
+            {
+                SWSS_LOG_ERROR("Failed to create ACL table EgressSetDSCP");
+                // since we failed to create the table, there is no need for rollback.
+                return EgressSetDscpTableStatus::EGRESS_SET_DSCP_TABLE_FAILED;
+            }
+        }
+        else
+        {
+            if (updateAclTable(m_AclTables[egrSetDscp_oid], egrSetDscpTable))
+            {
+                SWSS_LOG_INFO("Successfully updated existing ACL table EgressSetDSCP");
+                // We do not set the status here as we still have to update
+                // TABLE_TYPE_MARK_META/V6 table.
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Failed to update existing ACL table EgressSetDSCP");
+                //  there is no need for roollback as we have not made any changes to the MARK_META/V6 tables.
+                // We can simply return false.
+                return EgressSetDscpTableStatus::EGRESS_SET_DSCP_TABLE_FAILED;
+            }
+        }
+        // keep track of the fact that this table is now associated with the EGR_SET_DSCP table.
+
+        m_egrSetDscpRef.insert(table_id);
+        SWSS_LOG_INFO("Added ACL table %s to EgrSetDscpRef", table_id.c_str());
+        status = EgressSetDscpTableStatus::EGRESS_SET_DSCP_TABLE_SUCCESS;
+
+    }
+
+    return status;
+}
+
+bool AclOrch::removeEgrSetDscpTable(string table_id)
+{
+    m_egrSetDscpRef.erase(table_id);
+    SWSS_LOG_INFO("attempting to remove %s reference", table_id.c_str());
+    if (m_egrSetDscpRef.size() == 0)
+    {
+        if (!removeAclTable(EGR_SET_DSCP_TABLE_ID))
+        {
+            SWSS_LOG_INFO("Failed to remove the %s table so readding the reference of %s", EGR_SET_DSCP_TABLE_ID, table_id.c_str());
+            m_egrSetDscpRef.insert(table_id);
```

Comment:
Suggest removing the table from `m_egrSetDscpRef` if the removal is successful so that you don't need to add it back in failing case.

Reply:
the m_egrSetDscpRef is an internal structure which keeps track of the active tables which are using EGR_SET_DSCP_TABLE. we also use it during the port set calcualtions. when removing a table it has to be derefed form this dictionary first bacause we use this to find out what other tables are referencing the EGR_SET_DSCP_TABLE The only other solution is to create a copy, remove the table from the copy and send it down. I think that would overcomplecate the solution.